File: Backend/Microservices/Bursary/Eefa.Bursary.Application/UseCases/Definitions/BankBranch/Commands/Add/CreateBankBranchCommand.cs
using AutoMapper;
using Eefa.Bursary.Application.UseCases.Definitions.Bank.Commands.Add;
using Eefa.Bursary.Domain.Entities.Definitions;
using Eefa.Bursary.Infrastructure.Interfaces;
using Eefa.Common;
using Eefa.Common.CommandQuery;
using Eefa.Common.Exceptions;
using MediatR;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace Eefa.Bursary.Application.UseCases.Definitions.BankBranch.Commands.Add
{
    public class CreateBankBranchCommand : CommandBase, IRequest<ServiceResult<BankBranches>>, IMapFrom<CreateBankBranchCommand>, ICommand
    {
        public int BankId { get; set; } = default!;
        public string Code { get; set; } = default!;
        public string Title { get; set; } = default!;
        public int CountryDivisionId { get; set; } = default!;
        public string Address { get; set; } = default!;
        public string TelephoneJson { get; set; }
        public string ManagerFullName { get; set; }

        public void Mapping(Profile profile)
        {
            profile.CreateMap<CreateBankBranchCommand, BankBranches>().IgnoreAllNonExisting();
        }


    }

    public class CreateBankBranchCommandHandler : IRequestHandler<CreateBankBranchCommand, ServiceResult<BankBranches>>
    {

        private readonly IMapper _mapper;
        private readonly IBursaryUnitOfWork _uow;

        public CreateBankBranchCommandHandler(IMapper mapper, IBursaryUnitOfWork uow)
        {
            _mapper = mapper;
            _uow = uow;
        }

        public async Task<ServiceResult<BankBranches>> Handle(CreateBankBranchCommand request, CancellationToken cancellationToken)
        {
            var bankbranch = _mapper.Map<BankBranches>(request);

            if (bankbranch == null)
            {
                throw new ValidationError("اطلاعات ورودی ارسال نگردیده است");
            }
            _uow.BankBranches.Add(bankbranch);
            var value = await _uow.SaveChangesAsync(cancellationToken);

            if (value <= 0)
                throw new Exception("بروز خطا در ثبت اطلاعات بانک");

            return ServiceResult<BankBranches>.Success(bankbranch);
        }
    }


}

File: Frontend/angular-app/src/app/modules/bursary/repositories/bank-branch/commands/create-bank-branch-command.ts
import {Inject} from "@angular/core";
import {ServiceResult} from "../../../../../core/models/service-result";
import {IRequest, IRequestHandler} from "../../../../../core/services/mediator/interfaces";
import {HttpService} from "../../../../../core/services/http/http.service";
import {ApplicationError} from "../../../../../core/exceptions/application-error";
import {MediatorHandler} from "../../../../../core/services/mediator/decorator";
import {ValidationRule} from "../../../../../core/validation/validation-rule";
import {HttpRequest} from "../../../../../core/services/http/http-request";
import {NotificationService} from "../../../../../shared/services/notification/notification.service";
import {BankBranch} from "../../../entities/bank-branch";

export class CreateBankBranchCommand extends IRequest<CreateBankBranchCommand, BankBranch> {
  bankId:number | undefined = undefined;
  bankTitle:string | undefined = undefined;
  code:string | undefined = undefined;
  title:string | undefined = undefined;
  countryDivisionId:number | undefined = undefined;
  address:string | undefined = undefined;
  phoneNumber:string | undefined = undefined;
  managerFullName:string | undefined = undefined;


  constructor() {
    super();
  }

  mapFrom(entity: BankBranch): CreateBankBranchCommand {
    throw new ApplicationError(CreateBankBranchCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  mapTo(): BankBranch {
    throw new ApplicationError(CreateBankBranchCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  get url(): string {
    return "/bursary/bankBranches/add";
  }

  get validationRules(): ValidationRule[] {
    return [];
  }
}

@MediatorHandler(CreateBankBranchCommandHandler.name)
export class CreateBankBranchCommandHandler implements IRequestHandler<CreateBankBranchCommand, BankBranch> {
  constructor(
    @Inject(HttpService) private _httpService: HttpService,
    @Inject(NotificationService) private _notificationService: NotificationService
  ) {
  }

  async Handle(request: CreateBankBranchCommand): Promise<BankBranch> {
    let httpRequest: HttpRequest<CreateBankBranchCommand> = new HttpRequest<CreateBankBranchCommand>(request.url, request);
    // httpRequest.Query  += ;
    // httpRequest.Headers = httpRequest.Headers.append()

    return await this._httpService.Post<CreateBankBranchCommand, ServiceResult<BankBranch>>(httpRequest).toPromise().then(response => {
      this._notificationService.showSuccessMessage()
      return response.objResult
    })
  }
}
